course2/lecture_en.md

```diff
@@ -0,0 +1,379 @@
+    \mapsto & 5 * 3 && \text{because $5 - 2 \mapsto 3$} \\
+    \mapsto & 15 && \text{because $5 * 3 \mapsto 15$}
+\end{align}$$
+
+Therefore, $(2 + 3) * (5 - 2) \Rightarrow 15$.
+
+#### Variable Binding
+
+In MoonBit, variable binding can be achieved using the syntax `let <identifier> : <type> = <expression>`. It assigns an identifier to a value that is represented by an expression. In many cases, the type declaration is optional as the compiler can infer it based on the type of the expression.
+- `let x = 10`
+- `let y = "String"`
+
+Rebinding an identifier in MoonBit will result in shadowing the previous value associated with that identifier, rather than modifying it. This means that the new value assigned to the identifier will take precedence over the previous value within the scope where it is rebound.
+
+By utilizing variable binding effectively, one can avoid complex nesting of expressions and make their code more readable and maintainable.
+
+#### Expression Blocks and Scope
+
+In MoonBit, expression blocks can be defined using the syntax
+```
+{
+  Variable bindings
+  Variable bindings
+  ……
+  Expression
+}
+```
+
+The type/value of an expression block is the type/value of the last expression.
+
+When a function or identifier is defined outside of any expression block, it is called a top-level definition. Conversely, when a function or identifier is defined within an expression, it is referred to as a local definition.
+
+The terms "top-level" and "local" are used to describe the scope in which these definitions are effective. Top-level definitions have a global scope, meaning they are valid throughout the entire file, while local definitions have a limited scope, starting from the point of definition and ending at the completion of the enclosing expression block.
+
+![height:8cm](../pics/scope.drawio.png)
+
+For example, in the above figure, we have defined two top-level identifiers: `value` and `y`. The value of `value` is determined by the expression block, which includes the bindings of `x`, `tmp`, and `another_tmp`. In `another_tmp`, we have another binding of `tmp`.
+
+It is important to note the direction of the arrows. On line 7, the `tmp` refers to the most recent definition of `tmp` on line 5, overshadowing the definition on line 3. However, on line 9, `tmp` refers to the definition on line 3, as it is now outside the scope of the `tmp` defined on line 5.
+
+#### Expression Reduction under Variable Binding
+
+Expression reduction can be broken down into the following steps:
+- Reduce the expression on the right-hand side of the variable binding.
+- **Replace** occurrences of identifiers with their reduction results.
+- Omit the variable binding part.
+- Reduce the remaining expressions.
+
+Take the following code snippet as an example:
+
+```moonbit expr
+let y: Int = 10
+let value = {
+  let x = 1
+  let tmp = x * 2
+  let another_tmp = {
+    let tmp = x * 3
+
+    tmp
+  }
+  tmp + another_tmp + y
+}
+```
```

Comment:
I moved the definition of `y` above `value`. Therefore, I removed the following description in the English version:
> 同时我们注意到，第9行的`y`的定义出现在第12行，也就是定义出现在了使用之后。这是被允许的，因为`y`在顶层被定义，作用域为全局。

I have not synced the content in the image yet.

Reply:
I now think there is a problem with my original modification: when we marked the code as `moonbit expr`, it was treated as an expression within `fn init`, so strictly speaking `y` and `value` are no longer top-level, and there will be a mismatch between the code and the explanation.
Maybe I should keep the original code structure unchanged and just add a type annotation for `value`.